facebook/delphiFacebook/R/contingency_write.R

```diff
@@ -17,35 +20,168 @@
+write_contingency_tables <- function(data, params, geo_type, groupby_vars)
 {
   if (!is.null(data) && nrow(data) != 0) {
-    data <- arrange(data, across(all_of(groupby_vars)))
+    
+    # Reorder the group-by columns and sort the dataset by them.
+    groupby_vars <- c("geo_id", sort(setdiff(groupby_vars, "geo_id")))
+    data <- data %>%
+      select(all_of(groupby_vars), everything()) %>%
+      arrange(across(all_of(groupby_vars)))
 
     # Format reported columns.
-    data <- mutate_at(data, vars(-c(groupby_vars)),
-                      function(x) {
-                        stri_trim(
-                          formatC(as.numeric(x), digits=7, format="f", drop0trailing=TRUE)
-                        )
-                      })
-
-    # Reduce verbosity of grouping vars for output purposes
-    groupby_vars <- gsub("_", "", sub(
-      ".+?_", "", groupby_vars[groupby_vars != "geo_id"]))
-    filename <- sprintf("%s_%s.csv", format(params$start_date, "%Y%m%d"),
-                        paste(c(geo_level, groupby_vars), collapse="_"))
-    file_out <- file.path(params$export_dir, filename)
-
+    format_number <- function(x) {
+      stri_trim(formatC(as.numeric(x), digits=7, format="f", drop0trailing=TRUE))
+    }
+    data <- mutate_at(data, vars(-groupby_vars), format_number)
+    
+    # Add standard geographic and metadata variables to the data.
+    data <- add_geo_vars(data, params, geo_type)
+    data <- add_metadata_vars(data, params, geo_type, groupby_vars)
+    
     create_dir_not_exist(params$export_dir)
-
-    msg_df(sprintf("saving contingency table data to %-35s", filename), data)
-    write_csv(data, file_out)
+    
+    file_name <- get_file_name(params, geo_type, groupby_vars)
+    msg_df(sprintf("saving contingency table data to %-35s", file_name), data)
+    write_csv(data, file.path(params$export_dir, file_name))
 
   } else {
     msg_plain(sprintf(
       "no aggregations produced for grouping variables %s (%s); CSV will not be saved",
-      paste(groupby_vars, collapse=", "), geo_level
+      paste(groupby_vars, collapse=", "), geo_type
     ))
   }
 }
+
+#' Add geographic variables to a dataset, e.g. state and state FIPS codes.
+#' 
+#' @param data A data frame, containing the variables in groupby_vars.
+#' @param params A parameters object with the `static_dir` resources folder.
+#' @param geo_type "nation", "state".
+#' 
+#' @importFrom dplyr bind_cols left_join select
+#' @importFrom readr read_csv cols
+#' @noRd
+add_geo_vars <- function(data, params, geo_type) {
+  
+  start <- data.frame(
+    country = "United States",
+    ISO_3 = "USA",
+    GID_0 = "USA"
+  )
+  
+  if (geo_type == "nation") {
+    
+    rest <- data.frame(
+      region = "overall",
+      GID_1 = NA_character_,
+      state = "overall",
+      state_fips = NA_character_,
+      county = "overall",
+      county_fips = NA_character_
+    )
+    
+  } else if (geo_type == "state") {
```

Comment:
We want to support county and Metropolitan Statistical Area (MSA) levels here as well. CMU is not currently releasing any such aggregates, but we will in the (near, hopefully) future. County and MSA don't correspond to any geo-levels used in international tables, but we should use the same output format.

My conception of how this would work: Following the county/state precedent, we'd just add another couple columns, `msa` and `msa_id`, on to support MSA grouping. MSAs are made up of 1 or more counties and cross state lines, so when aggregations use MSAs, states and counties should be set to "overall"/"all".

There's an existing [static file that lists MSA](https://github.com/cmu-delphi/covidcast-indicators/blob/main/facebook/static/msa_list.csv) names and corresponding CBSA ids that can be used.

Reply:
This change is larger and a "nice to have"; If it takes too long to add, we can add in a separate PR later on.

Reply:
I would like to leave this one till later since I don't have the time to address it right now.